### ISSUE TITLE
Unable to declare NFS mountpoint for LXC

### ISSUE THREAD

**Issue by GiovanniBattista** [Telmate/terraform-provider-proxmox#277]
I have a NFS share mounted on the Proxmox host, which I want to get mounted inside an LXC. 
Therefore, I have created the following terraform file. I couldn't find much information about how to specify a "mountpoint" (e.g. what is "key" or "slot"), therefore I guessed much of the values (although I have tried to match the keys with [this](https://pve.proxmox.com/wiki/Linux_Container#_bind_mount_points)).

**main.tf**
```
terraform {
  required_providers {
    proxmox = {
      source = "Telmate/proxmox"
      version = "2.6.5"
    }
  }
}
provider "proxmox" {
  pm_api_url = "<URL>"
  pm_user = "<USER>"
  pm_password = "<PWD>"
  pm_tls_insecure = "true"
}
resource "proxmox_lxc" "plex" {
  hostname = "plex"
  target_node = "pve"

  cores = 1
  memory = 2048

  ostemplate = "local:vztmpl/ubuntu-20.10-standard_20.10-1_amd64.tar.gz"
  ostype = "ubuntu"
  unprivileged = "true"
  start = "true"

  rootfs {
    storage = "os_drive_mirror"
    size = "10G"
  }
  swap = 0

  mountpoint {
    key = "1"
    slot = 0
    storage = "/mnt/media"
    mp = "/mnt/media"
    size = "100G" #this is required
  }

  network {
    name ="eth0"
    bridge = "vmbr0"
    # firewall = "true" # this does not work at the moment
    ip = "10.0.0.80/24"
    gw = "10.0.0.138"
  }

  ssh_public_keys = file("~/.ssh/id_rsa.pub")
}
```

Without the "mountpoint" part, the container gets created within Proxmox. But with "mountpoint" enabled, `terraform apply` throws the following error:
```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # proxmox_lxc.plex will be created
  + resource "proxmox_lxc" "plex" {
      + arch            = "amd64"
      + cmode           = "tty"
      + console         = true
      + cores           = 1
      + cpulimit        = 0
      + cpuunits        = 1024
      + hostname        = "plex"
      + id              = (known after apply)
      + memory          = 2048
      + onboot          = false
      + ostemplate      = "local:vztmpl/ubuntu-20.10-standard_20.10-1_amd64.tar.gz"
      + ostype          = "ubuntu"
      + protection      = false
      + ssh_public_keys = <<-EOT
            ssh-rsa <SSH KEY>
        EOT
      + start           = true
      + swap            = 0
      + target_node     = "pve"
      + tty             = 2
      + unprivileged    = true
      + vmid            = 0

      + mountpoint {
          + acl       = false
          + backup    = false
          + key       = "1"
          + mp        = "/mnt/media"
          + quota     = false
          + replicate = false
          + shared    = false
          + size      = "100G"
          + slot      = 0
          + storage   = "/mnt/media"
          + volume    = (known after apply)
        }

      + network {
          + bridge = "vmbr0"
          + gw     = "10.0.0.138"
          + hwaddr = (known after apply)
          + ip     = "10.0.0.80/24"
          + name   = "eth0"
          + tag    = (known after apply)
          + trunks = (known after apply)
          + type   = (known after apply)
        }

      + rootfs {
          + size    = "10G"
          + storage = "os_drive_mirror"
          + volume  = (known after apply)
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

proxmox_lxc.plex: Creating...

Error: Error creating LXC container: unable to create CT 106 - directory '/mnt/media:100' does not exist, error status:  
(params: {"arch":"amd64","cmode":"tty","console":true,"cores":1,"cpulimit":0,"cpuunits":1024,"features":"","hostname":"plex","memory":2048,"mp0":"/mnt/media:100,mp=/mnt/media","net0":"name=eth0,ip=10.0.0.80/24,bridge=vmbr0,gw=10.0.0.138","onboot":false,"ostemplate":"local:vztmpl/ubuntu-20.10-standard_20.10-1_amd64.tar.gz","ostype":"ubuntu","protection":false,"rootfs":"os_drive_mirror:10","ssh-public-keys":"ssh-rsa <SSH KEY>\n","start":true,"storage":"local","swap":0,"tags":"","tty":2,"unprivileged":true,"vmid":106})
```
IMO, "size" does not make sense in this context, and therefore should not be mandatory?
Is this an issue or should I specify something differently?
Maybe also update the docu with some examples for different mountpoints (storage backed vs. bind mount

**Comment by snoord** [comment]
So I was in the process of updating the LXC resource documentation and after going through the code and the PVE docs, I think I have discovered what is causing this.

Proxmox supports a few different types of mount points for LXC containers - they are [Storage Backed](https://pve.proxmox.com/pve-docs/pve-admin-guide.html#_storage_backed_mount_points), [Bind](https://pve.proxmox.com/pve-docs/pve-admin-guide.html#_bind_mount_points), and [Device](https://pve.proxmox.com/pve-docs/pve-admin-guide.html#_device_mount_points) mount points. I found the following quote to be particularly relevant to this issue with LXC mount points:

> The special option syntax `STORAGE_ID:SIZE_IN_GB` for storage backed mount point volumes will automatically allocate a volume of the specified size on the specified storage. For example, calling
> ```
> pct set 100 -mp0 thin1:10,mp=/path/in/container
> ```
> will allocate a 10GB volume on the storage `thin1` and replace the volume ID place holder `10` with the allocated volume ID, and setup the moutpoint in the container at `/path/in/container`

While I can't say exactly where this is happening (`terraform-provider-proxmox`, `proxmox-api-go`, or the actual Proxmox API itself), it looks like the `storage` and `size` attributes of `mountpoint` are being concatenated (without the unit suffix from `size`) to form a string that emulates the `STORAGE_ID:SIZE_IN_GB` syntax.

Using your example @GiovanniBattista, this is why:
```hcl
  mountpoint {
    key     = "1"
    slot    = 0
    storage = "/mnt/media"
    mp      = "/mnt/media"
    size    = "100G"
  }
```

...results in this part of the error message:
```
[...] directory '/mnt/media:100' does not exist [...]
```

This concatenation looks to be assigned as the value for the `mountpoint.volume` attribute, which I believe is the case because manually defining the `mountpoint.volume` attribute like in the following example appears to be successful:
```hcl
  mountpoint {
    key     = "1"
    slot    = 0
    storage = "/mnt/media"
    mp      = "/mnt/media"
    size    = "100G"
    volume  = "/mnt/media" # <-- new
  }
```

Now with all that said, after creating an LXC container with the above `mountpoint` block, I am seeing the following error when using `terraform plan`:

```
Error: Proxmox Provider Error: proxmox API returned new parameter 'file' we cannot process
```

So there is still work that needs to be done, but I hope this was at least helpful in the short-term. I will also be opening a pull request with improved documentation for the LXC resource, though it still needs a tiny bit of work before it is ready :slightly_smiling_face:

**Comment by tiehfood** [comment]
I'm also struggling with this. If I create an LXC-Container and add a bind-mount by hand, I would normally add something like this to the CONTAINER-ID.conf file:
```
mp0: /mnt/share,mp=/mnt/share
```
so there should be an option in terraform to declare the type of the mountpoint, which defines the mandatory attributes. In my case only the _key_, _mp_ and _volume_ attributes would be needed

**Comment by frozenfoxx** [comment]
I'm having this same problem. I don't need it to create a storage-backed volume, I need it to create a bind mount for passthrough of a NFS location on the host. Without `size` set to optional this seems impossible.

**Comment by arawak** [comment]
I'd also love to see this. Currently I have to manually add 

```
mp0: /data/docker,mp=/storage
```

to the .conf file or use ansible to do it.

**Comment by jtorrex** [comment]
Does anyone have an example of how to bind mount host paths inside LXCs?

**Comment by zinefer** [comment]
I am upgrading from an admittedly very old version of this provider and proxmox itself and am getting hung up on `size` seemingly being required now. On the version I was using prior it was not required and I have an entire cluster of LXC containers that have been set up through the (admittedly older version) proxmox api without specifying a `size`.

**Comment by zinefer** [comment]
As far as I know this is still an issue. There is a proposed solution above that has yet to receive review.

**Comment by zinefer** [comment]
This issue ended up being resolved by merging #844. Thanks everyone!
